rfcs/20201014-VTune-ITT-tagging/README.md

```diff
@@ -0,0 +1,142 @@
+Users can understand performance issues among each oneDNN primitives.  
+![](images/group.JPG "group by oneDNN primitives")
+
+### 3. map primitive with related computation kernels.  
+By tagging primitives, users can also map primitive with computation kernels.  
+For example, users can understand one inner primitive contains several gemm_nocopy_f32 kernels.  
+![](images/ops_mapping.JPG "mapping between oneDNN primitives and kernels")
+
+## Proposal
+
+Add VTune ITT tasks for all primitives.
+Users can see profiling results grouped by primitives. 
+primitive grouping is more useful than function grouping to DL developers.
+### Implementation
+To enable VTune ITT tagging, we includes ittnotify related source files in projects https://github.com/intel/ittapi under src/common/ittnofiy folder.  
+We also create ittnotify.cpp and ittnotify.hpp under /src/common folder to use ITT functions from copied ITT source files.
+We introduce a new namespace itt under dnnl namespace.
+In the same sapce we have 4 DNNL_API for using ITT tagging.
+
+| functions  | Note |  
+|---------------------|---------------------|  
+| void DNNL_API primitive_task_start(primitive_kind_t kind) |  use this function to tag start of a ITT task with primitive kind info  | 
+| void DNNL_API primitive_task_end()  |    use this function to tag end of a ITT task    | 
+| primitive_kind_t DNNL_API primitive_task_get_current_kind() |   get the primitive task for current ITT task       | 
+| bool DNNL_API get_itt(__itt_task_level level) |    get ITT level.  enable different ITT tagging by using this level variable  |   
+
+We also define a enum type with different itt task level.  
+Here is a table for meaning of all itt task levels.  
+
+| __itt_task_level  | Note |  
+|---------------------|---------------------|  
+| __itt_task_level_none |   no ITT event will be triggered  | 
+| __itt_task_level_low  |    ITT events are only triggered in master thread     | 
+| __itt_task_level_high |    ITT events are triggered in all OMP/TBB threads      |   
+
+There is a new struct itt_task_level_t with only one int variable in it.  
+~~~cpp
+struct itt_task_level_t {
+    int level;
+};
+~~~
+
+#### Implementation details :  tag primitive execution with ITT
+---
+Here are pesudo codes for how we define primitive execution on master thread as a ITT task.
+~~~cpp
+status_t dnnl_primitive_execute(const primitive_iface_t *primitive_iface,
+    ....
+    if (itt::get_itt(itt::__itt_task_level_low))
+        itt::primitive_task_start(
+            primitive_iface->pd()->impl()->kind());
+
+    .... // Primitive Execution
+
+    if (itt::get_itt(itt::__itt_task_level_low))
+        itt::primitive_task_end();
+    return status;
+}
+
+~~~
+
+### Feature Enablement
+---
+**Compilation Time**  
+
+We enable this feature on compilation time by default.  
+User can disable the feature by following CMake declaration at compiling time.    
+DNNL_ITT_TASKS=False   
```

Comment:
Does Pytorch (or other frameworks) have to be built with extra compilation flags (such as debug symbols) in order to perform performance profiling? If recompilation is required, then I think these flags should be disabled by default on oneDNN side. This will also work well with the case when the user wishes to provide a custom path for vtune.

Reply:
Besides, I think `DNNL_ITT_TASKS` should be renamed to something closer to vtune, perhaps `DNNL_ENABLE_VTUNE_TAGS` or `DNNL_ENABLE_VTUNE_TASKS`

Reply:
I changed it accordingly.
if it works for all, I will also make the changes in MR.
https://gitlab.devtools.intel.com/ipl/dl/mkl-dnn/-/merge_requests/2091

Reply:
Feedback from VTune team : "Since ITT API was open sourced it is != VTune anymore. Anyone can write ITT collector and benefit from this instrumentation – and we are working on publishing reference implementation of such collector at the moment. So I would position this change not as something special for VTune but a thing generally useful for perf analysis of AI workloads running on top of oneAPI  "

To address the feedback from VTune team, I plan to replace "VTune" with ITT.
Let me know any concern for that plan.

Reply:
That sounds very reasonable to me, thanks for following up with feedback from VTune folks.